### ISSUE TITLE
Bounded iteration

### ISSUE THREAD

**Issue by llogiq** [llogiq/mutagen#23]
Apart from timeouts (which aren't implemented yet either) which the mutation test runner needs to take care of, we should probably introduce bounds for all loops we encounter. This means we issue a loop ID for each iteration, and modify each loop to setup a "loop run" object that records the number of iterations (for the baseline test run, i.e. mutation_count == 0). On drop, the loop run object compares its iteration count with the maximum loop count for the loop ID and if higher, updates the maximum loop count.

Then on mutation runs, the object compares the current loop count with *a base value* `+` *the recorded maximum* `*` *some factor* and if higher, panics, thus limiting test runtime without needing to kill the process (on the other hand, we may want to kill it nevertheless, because a failing test may leave the process in a detrimental state). Even then, we allocate less runtime for a test that is very likely to fail.

**Comment by gnieto** [comment]
@llogiq I've started to work on this . I remember that on some issue or PR you mentioned that you got some idea on how to do the Interprocess Communication.  What you had in mind for this? I've been checking some options, but there aren't many options with Windows support. I was thinking on using sockets, but it may get complex. Do you have any idea or preference?

Besides of the IPC thing, I guess that we may do something like:

1) Run the tests without mutations tracing the max iterations for a mutated loop
   - From the tests, send the max bounded iteration to the runner. I guess that we will need to do this communication or at the end of any loop (easy and inefficient) or at the end of the mutated function/method (a little bit harder but more efficient).
   - The runner should save the max bound of all the loops
2) Launch the mutated tests
   - The runner needs to communicate to the tests the max iteration for each of the loops. I guess we may use something like lazy_static to initialize this information when we execute the first loop on mutated code.

I'm missing something?

**Comment by llogiq** [comment]
Funny thing – I was just writing up my preferred solution for my blog, and it is indeed method-wide. The trick is to have a new type LocalCount containing the (local) counter and a ref to the static AtomicUsize counter. 

When `mutation_count == 0`, we set up this local counter, it gets `drop`ped on method exit. It should also be reset at the start of the corresponding loop.

On`drop`, we can use the `AtomicUsize::fetch_max` method I just got into libcore to update the global count and if necessary send the update back to the runner (e.g. via file or else).

On `mutation_count > 0` we can use a lazy_static as per your suggestion.
